File: 0140-word-break-ii/0140-word-break-ii.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        def wordsEndingIn(i):
            if i == len(s):
                return [""]
            ans = []
            for j in range(i+1, len(s)+1):
                if s[i:j] in wordDict:
                    for tail in wordsEndingIn(j):
                        if tail != '':
                            ans.append(s[i:j] + " " + tail) 
                        else:
                            ans.append(s[i:j])
            return ans
        return wordsEndingIn(0)

File: 0140-word-break-ii/0140-word-break-ii.java
class Solution {
    private HashSet<String> set;

    public List<String> wordBreak(String s, List<String> wordDict) {
        this.set = new HashSet<>(wordDict);
        List<String> words = generateValidWords(s, 0);
        return words;
    }

    private List<String> generateValidWords(String s, int index) {
        if (index == s.length()) {
            return Collections.emptyList();
        } else {
            List<String> words = new ArrayList<>();
            for (int nextIndex = index + 1; nextIndex <= s.length(); ++nextIndex) {
                String prefix = s.substring(index, nextIndex);
                if (set.contains(prefix)) {
                    String suffixString = s.substring(nextIndex);
                    List<String> suffixes = generateValidWords(suffixString, 0);
                    if (suffixes.isEmpty()) {
                        if (suffixString.isEmpty()) {
                            words.add(prefix);
                        }
                    } else {
                        for (String suffix : suffixes) {
                            words.add(prefix + " " + suffix);
                        }
                    }
                }
            }
            return words;
        }
    }
}